### ISSUE TITLE
Support Redis as caching backend

### ISSUE THREAD

**Issue by CAFxX** [pagespeed/ngx_pagespeed#777]
ngx_pagespeed does not currently allow to use Redis as a cache backend: are there any plans to support it?

**Comment by Skoks** [comment]
Any updates ?

**Comment by jeffkaufman** [comment]
No one has been working on this, and it's not currently on anyone's schedule.

If someone is looking to get into PageSpeed development this would be a great bug to work on!

**Comment by jeffkaufman** [comment]
Added to [Accepted Feature Requests](https://github.com/pagespeed/mod_pagespeed/wiki/Accepted-Feature-Requests).

**Comment by mariarti** [comment]
+1

**Comment by f4rsh4d** [comment]
+1

**Comment by josuegio** [comment]
+1

**Comment by A-David** [comment]
Any updates?

**Comment by jeffkaufman** [comment]
It looks like this will probably be an intern project for this summer.

**Comment by xjunior** [comment]
Thanks for the response, @jeffkaufman

**Comment by isaumya** [comment]
+1 Looking forward to this fix

**Comment by akincansenol** [comment]
I am waiting impatiently for!

**Comment by gamerjota** [comment]
Looking forward for this! Imagine the impact of this two working together.

**Comment by isaumya** [comment]
Any update yet?

**Comment by jeffkaufman** [comment]
We have an intern who's going to work on this starting about three months from today.

**Comment by f4rsh4d** [comment]
It would be great if the focus will be on Redis Cluster , a single pipe caching mechanism can be achieved by Memcached for now and it's ok until we use two or more server and in case one of them get out of service Pagepseed don't handle the situation nice  and in some areas it takes time for a page to load due to inability to connect to a certain Memcached server.

I believe if Redis is going to be supported Redis Cluster should be in mind beside normal Redis instance.

**Comment by nathanjosiah** [comment]
It has been nearly a month since this feature was reported to be worked on, is there any update or ETA @jeffkaufman?

**Comment by yeputons** [comment]
@nathanjosiah I'm the intern working on that. First commit has already landed <a href="https://github.com/pagespeed/mod_pagespeed/commit/0b7fb5e12a43bc04b515cb640ffca839ec4542e1">in mod_pagespeed</a>, but there is still a lot of work - e.g. Redis adapter is not integrated with the rest of code yet.

I will update here whenever master branch of mod_pagespeed has something minimal that you can actually play with.

**Comment by nathanjosiah** [comment]
Thank you for this update @yeputons! Are you thinking this will land within weeks or months?

**Comment by yeputons** [comment]
@nathanjosiah I cannot promise anything, but I expect it to land within a month or so.

**Comment by isaumya** [comment]
Thanks a lot @yeputons for working on this issue. Really huge thanks. Last time I checked, every time I turned on Redis Cache, pagespeed stopped working. I am using @centminmod on my server. I hope after you add the redis feature into pagespeed I can use both of them simultaneously. :)

**Comment by jmarantz** [comment]
FWIW I thought I'd paste a few searches showing memcached's implementation in mod/ngx_pagespeed so folks understand the amount of work required for this:

https://github.com/pagespeed/mod_pagespeed/search?utf8=%E2%9C%93&q=memcache&type=Code
https://github.com/pagespeed/mod_pagespeed/search?utf8=%E2%9C%93&q=memcached&type=Code
https://github.com/pagespeed/mod_pagespeed/search?utf8=%E2%9C%93&q=AprMemCache&type=Code

I don't know how to express that with one search, and I'm not sure those three searches capture the depth of the integration effort :)


I'm curious, though, if someone can articulate why they prefer redis over memcached for mod_pagespeed's use.  I just skimmed through http://www.infoworld.com/article/3063161/application-development/why-redis-beats-memcached-for-caching.html which is obviously pro-redis .  However I think to exploit redis' advantages, mod_pagespeed would need to have much more extensive changes about how it interacts with caches, which would extend way beyond an intern project.

Of course, a perfectly good reason might just be "our site cache everything else with redis, so we don't want to add memcached just for mod_pagespeed".   But I am not catching what metrics would improve by replacing mod_pagespeed/memcached with mod_pagespeed/redis.

**Comment by jmarantz** [comment]
@isaumya  can you give more detail on what failed when you ran mod_pagespeed and redis at the same time?  Since as of today mod_pagespeed is not integrated to redis, I'm not sure whether you are describing a functional problem with your net stack and caching at various layers (e.g. Redis caches unoptimized HTML), or whether it's a resource usage issue where combining MPS & Redis on one physical machine loads it down too much.  

If it's a caching semantics issue, you might want to look at https://developers.google.com/speed/pagespeed/module/downstream-caching

**Comment by isaumya** [comment]
@jmarantz Hi. Thanks for yur reply. I use nginx_pagespeed not mod_pagespeed. As I stated above I use @centminmod (http://centminmod.com/) on my server. Now centminmod has an inbuilt option to enable redis cache.
But every time I enable it I see the css/js combination/minification, image optimization, webp conversion is not working anymore. I just see the normal HTML without any pagespeed magic.

**Comment by jmarantz** [comment]
I see...I think the Redis integration we are implementing right now might be somewhat orthogonal to your particular issue, which appears related to https://developers.google.com/speed/pagespeed/module/downstream-caching -- please take a look at that :)

I'm curious whether other people on this thread are expecting to get out of a Redis integration.

**Comment by JoyceBabu** [comment]
> Of course, a perfectly good reason might just be "our site cache everything else with redis, so we don't want to add memcached just for mod_pagespeed".

I am in that group. Currently I am using both memcached and Redis. Memcached is used only by MPS. If I can replace Memcached with Redis, I would prefer that, as long as there is no performance degradation.

> I think maybe one compelling advantage is checkpointing the redis contents to disk on shutdown, so that machine reboot != cache flush

I faced this problem today. My memcached server was up for the past 30 days, and I was seeing a large number of cache evictions per second (around 10/s). So, I decided to raise the cache size to 8GB and searched for a solution to do that without clearing the cache. The only only option I could find was to use `memdump` to dump the keys and `memcat` to dump the values. Since I was worried that the cache might get corrupted during the export/import and it might end up serving corrupted files, I decided not to go for it. So, I went ahead and restarted the server without exporting, and watched the 6 million cached items vanish 😒 . If it was Redis, I could have used `SAVE` command to save the data to disk.

**Comment by justbeez** [comment]
> I'm curious, though, if someone can articulate why they prefer redis over memcached for mod_pagespeed's use.

@jmarantz TLDR; when used like Memcached, Redis is as fast (faster if you ask some)—but it also has a slew of other features that are important to many use cases, and some of which have big out-of-the-box benefits to *_pagespeed.

For us, obviously it's nice just because we're already using Redis (who wants to deal with both, especially when you're paying to make them HA), but I think this comes down much more to _why_ we're already using Redis over Memcached.  That breaks down into a few areas:

1. The ability to persist to disk (as already mentioned).  Don't need this in your pagespeed use case?  Great; turn it off and get a performance benefit.
2. Better support for horizontal scaling (including ease of setting up HA/cross-region failover, which is an out of the box feature through clustering).
3. More control over eviction policies (Memcached's LRU may very well discard an expensive object just because it's a similar size or it just feels like it).  This added control could also allow more robust purging to be integrated somewhat transparently.

There's a lot more that you _can_ benefit from, like optimistic locking in transactions, data larger than 1MB per key, simpler administration and reporting . . . there's a ton of features that just aren't present in the Memcached universe, or require you to bolt on a third-party solution.  When it comes down to it, Redis gives me a ton of options and puts me in control of what's best for my workflows.

(Reminder to anyone reading this, and before anyone jumps on the "specialization" bandwagon: every feature has a cost—and if you turn on all the bells and whistles in Redis, it's going to be slower than Memcached. If you're the sysadmin, be responsible and make sure the benefits outweigh the costs in **your** environment.)

**Comment by yeputons** [comment]
Hi everyone, work is in progress. We're currently discussing what features of Redis sharding/replication we want to support and we need your opinion on what you're going to use and why.

I believe it's possible to configure Redis in following four ways:
1. Single server, no sharding, no replication.
2. One master, multiple slaves - no sharding, but replication is available. However, it's not yet clear how exactly we will want to exploit that, because all writes should go to master if we want them to be visible.  We do not do any hard-to-process read requests. Another potential problem is having some delay between writing and being able to read that change from slave.
3. One master, multiple slaves (same as above), plus <a href="redis.io/topics/sentinel">Redis Sentinel</a> for failover.
4. <a href="http://redis.io/topics/cluster-tutorial">Redis Cluster</a> - solution for both sharding and replication.

We definitely want to support no. 1 as the "basic" configuration and we have seen requests for no. 4. Is there anyone who is going to use no. 2 or 3? If yes, why you do not use Redis Cluster instead?

It also looks to me that Redis Cluster is superior to Redis Sentinels, can anyone confirm/disprove that?

**Comment by fcorriga** [comment]
Redis cluster is more reliable than master/slave solutions, plus reads/wrires are in sync.
IMHO, 2 and 3 should be avoided at all costs, since there are write delays.
However, you could always start supporting 1 and 4 and add the other solutions if anyone needs master/slave setups.

**Comment by f4rsh4d** [comment]
For a high traffic website normally redis cluster for session and cache is the best solution , it would be great if pagespeed supports clustering too , for a low traffic or a single web server which normally redis resides inside the server , no1 is the choice.
I do agree with @fcorriga for 1 & 4.

**Comment by mariarti** [comment]
+ for 1 & 4

**Comment by justbeez** [comment]
Thanks again for the work being done here, and for the feedback everyone chipping in with!

I agree with focusing support on cases 1 and 4 (since they're by far the most common I've seen), and then address any other cases in the future if/when the community is able to articulate the need and help define reasonable implementation details.

Setting up HA Redis is complex, and most people I know are using cloud providers or manged deployments for it—so it may be wise to test the implementation using popular options like [Redis Labs](https://redislabs.com/redis-cloud). They have a free tier, and the original creator of Redis is in charge of their engineering—so that would be easy for testing case 1 and we might be able to get some engagement from them if needed. (I've also looked into Redis options from ObjectRocket, RedisToGo, ElastiCache, Azure, Heroku, and Compose.io)

Redis Labs' paid plans add some special sauce onto Redis Cluster, but it's all API-compatible.  We're currently using Redis Cloud for everything else in our setup—and I'm happy to test using our paid account as well. (We're actually on their Memcached Cloud for mod_pagespeed currently, so we'll be able to easily compare performance once the Redis feature is ready.)

**Comment by yeputons** [comment]
Hello everyone!

Current progress so far: `ExperimentalRedisServer` option became available in [mod_pagespeed@e53bc94](https://github.com/pagespeed/mod_pagespeed/commit/e53bc946e1c92c85b599048d26eb3976df7caef3). It passes all our unit and system tests, so it probably does not break everything immediately. Note that it's `Experimental` for a reason and it **is not ready for production** yet. Say, there are no timeouts on individual cache operations yet, so if network drops some packages, webserver will hang, which is no good. Also the thing does not pass our load tests (maybe it's Redis misconfiguration, maybe it's a bug).

Stay tuned, I keep working and I'm going to make another update next Friday!

**Comment by centminmod** [comment]
@yeputons thanks for the update and hard work on this, very much appreciated !

**Comment by yeputons** [comment]
Update: timeouts on individual cache operations are almost in (they work and are tested, almost all code review comments are addressed), load tests pass (looks like default Redis is slower than default Memcache: something like 4000qps vs 3500qps on some benchmark that was found nearby :), I've got a bunch of small follow-ups to do (regarding better multi-threading, error detection and not spamming logs when something goes wrong).

I hope to finish with all that functionality and basic speed testing by the end of next week. My next priority after that will be Redis Cluster.

**Comment by f4rsh4d** [comment]
@yeputons Any guess when we can test pagespeed and redis? 
you mentioned earlier after single instance you'll go for redis cluster , any update about that would appreciated also .

**Comment by yeputons** [comment]
@f4rsh4d my internship ends this Friday. Status so far:
* Single instance Redis should work, I've got timeouts and nice thread management in. Major missing things are MultiGet (which is a performance boost) and documentation. You can already experiment with it on master branch (setting is called `ExperimentalRedisServer` until we want to release it and stabilize configuration format). It should work, but it's still in "experimental" state.
* Redis Cluster is under heavy development. Most probably it won't be fully ready and tested by me till this Friday, but I'm hoping to check in most of the required stuff, so it will be fairly straightforward to polish.

I think @jefftk can elaborate more on our plans to release this feature (at least, single-instance version).

**Comment by f4rsh4d** [comment]
@yeputons Thanks for all the hard work during this summer , i'll try and test it ASAP.

**Comment by jeffkaufman** [comment]
The goal is for this to go out in the next major release, which we're planning to cut in ~2 weeks.  It should have single instance redis support, and will ideally have cluster support as well.

**Comment by Lofesa** [comment]
Hi to all
First of all, thanks for you great work.
@jeffkaufman and @yeputons , can i suggest the use of unix sockets to connect to redis when in standalone ?
some like this (php):
if ( self::$redis_port == 0) {
$connect = $redis->connect( self::$redis_host);
} else {
$connect = $redis->connect( self::$redis_host, self::$redis_port );
}
thx in advance

**Comment by jeffkaufman** [comment]
> can i suggest the use of unix sockets to connect to redis

People interested in unix sockets should follow https://github.com/pagespeed/mod_pagespeed/issues/760

We're not likely to have that ready for the first version.

**Comment by jeffkaufman** [comment]
(plan for today: stress-testing and benchmarking the single-instance redis support.)

**Comment by jeffkaufman** [comment]
Initial benchmarking results.  Apache configured with:

```
<VirtualHost localhost:8083>
  ServerName redis-stress.example.com

  DocumentRoot "path/to/mod_pagespeed_example"
  ModPagespeedFileCachePath "path/to/pagespeed-cache-redis"

  ModPagespeedExperimentalRedisServer localhost:6379

  ModPagespeedLoadFromFile "http://redis-stress.example.com/" \
    "path/to/mod_pagespeed_example/"
</VirtualHost>

<VirtualHost localhost:8083>
  ServerName memcached-stress.example.com

  DocumentRoot "path/to/mod_pagespeed_example"
  ModPagespeedFileCachePath "path/to/pagespeed-cache-memcached"

  ModPagespeedMemcachedServers localhost:6378

  ModPagespeedLoadFromFile "http://memcached-stress.example.com/" \
    "path/to/mod_pagespeed_example/"
</VirtualHost>

<VirtualHost localhost:8083>
  ServerName file-stress.example.com

  DocumentRoot "path/to/mod_pagespeed_example"
  ModPagespeedFileCachePath "path/to/pagespeed-cache-file"

  ModPagespeedLoadFromFile "http://file-stress.example.com/" \
    "path/to/mod_pagespeed_example/"
</VirtualHost>
```

siege.conf with added:
```
proxy-host = localhost
proxy-port = 8083
```

All runs interleaved: rotating round robin between redis, memcached, and file cache.

caches started with:
```
redis-server --port 6379 --maxmemory 64mb --maxmemory-policy allkeys-lru
memcached -m 64 -p 6378 -u memcache -l 127.0.0.1
```

Benchmarking cache reading with:

```
inititial: manual fetch of resource until optimized
siege http://[redis/memcached/file]-stress.example.com/[resource] --benchmark -t60s -c 50
```

Testing small reads with `rewrite_javascript.js`, which is 43 bytes after optimization (plus headers and things, so probably more like 400 bytes.  All number are trans/sec:

| redis | memcached | file |
| ---: | ---: | ---: |
|3336.13|3739.67|3457.31|
|3481.73|3531.41|3310.50|
|3447.95|3661.25|3248.83|
|3517.29|3475.72|3300.32|
|3397.41|3626.58|3343.16|
|3492.55|3574.33|3402.98|
|3369.77|3698.42|3316.22|
|3516.39|3527.76|3281.90|
|3378.46|3627.90|3267.39|
|3484.45|3618.63|3414.02|

Averages, and as a percentage of file-cache (default) performance:

* redis: 3442 (103%)
* memcached: 3608 (108%)
* file: 3334 (100%)

And here's a chart:

![chart for rewrite_javascript.js](https://cloud.githubusercontent.com/assets/262566/18714844/10532c1e-7fe5-11e6-9334-8402f8d72360.png)

Next I'll repeat this for a large file.

**Comment by jeffkaufman** [comment]
The performance difference between redis and memcached here is probably due to memcached having multiget support hooked up and redis not.

**Comment by f4rsh4d** [comment]
it's weird , in most scenario we had better performance with redis over memcached , but it was the time we used redis as a caching mechanism for mysql .
Is there any further optimization regarding redis instance ? i noticed you used 64 mb for both memcached and redis , what happens if this memory amount get raised to 256 or 512 or even 1gb .
Will you please benchmark with one or two of these values ?

**Comment by fcorriga** [comment]
There is a caveat. Single instance of Redis is single threaded while memcached is multitheraded

**Comment by fcorriga** [comment]
also... performance is not SO important when it's so close (however redis cluster is able to use multicore CPUs), versatility is the primary point.
Redis wins in data persistence, selective key deletion/eviction/expiration, datatype selection and now, with redis 3, ability to cluster (scalability) without external software. As a bonus redis has dynamic memory allocation thru jemalloc, so it doesn't waster you RAM if it does not need it.

**Comment by jmarantz** [comment]
I want to point out that the lack of MultiGet support is a temporary limitation of pagespeed's redis interface, not a limitation of redis.  The PagesSpeed team will work on Redis MultiGet in the coming weeks, but it's great to get a benchmark of where we are right now.

I also want to point out that a fundamental advantage of Redis is that it can back up to a file, which memcached cannot.  And while the benchmark above shows no speed advantage on this benchmark of Redis over file-cache, there might be one with a broader benchmark incorporating more writes to the DB -- I think Jeff's benchmark won't do many writes.

RE cache size: I don't think this matters for this particular benchmark, where there will only be one resource.  The cache could probably be 640k instead of 64M and still get the same results on this siege.

**Comment by jeffkaufman** [comment]
Testing large reads with `images/Beach.jpg`, which is 825,511 bytes optimized (and then some more bytes of header overhead).

|redis|memcached|file|
| ---: | ---: | ---: |
|742.41|737.10|743.86|
|728.06|741.01|724.74|
|739.85|732.64|722.37|
|742.90|734.54|716.46|
|740.43|737.15|724.77|
|727.60|727.66|733.83|
|735.56|725.18|719.61|
|728.74|723.02|736.91|
|739.10|725.26|756.67|
|763.79|755.00|755.54|

Averages:

* redis: 739 (101%)
* memcached: 734 (100%)
* file: 733 (100%)

![large-image-file-graph](https://cloud.githubusercontent.com/assets/262566/18716326/6df0359c-7fea-11e6-859d-d85f6f343013.png)

(Note: for even larger files, over 1MB, the memcache integration falls back to the file cache while the redis integration continues to use redis.)

**Comment by f4rsh4d** [comment]
@fcorriga for a single redis instance the performance is important not all the things you've mentioned are for redis as a caching mechanism.
redis cluster is intended to use while we have many webservers for an application and a single remote memcached can cause trouble and be a bottleneck if goes down , redis cluster provide HA and it's an important factor while it brings performance .

**Comment by fcorriga** [comment]
@f4rsh4d i perfectly agree. Also remember you can spawn multiple redis instances for several things. For example, we use 1 instance for php sessions, 1 instance for magento sessions, 1 instance for magento cache, 1 instance for magento full page cache. For example you can use a redis instance for each different (sub)domain with pagespeed enabled. You're not bound to a single redis instance and... best of all, you can cluster all those instances when redis cluster will be implemented in ngx_pagespeed.

**Comment by jeffkaufman** [comment]
@fcorriga "the file is small enough to fit in the filesystem cache. please, use a much larger number of different size files to replicate a real world scenario"

What I'm testing in this particular case is how fast the pagespeed-redis interface is, compared to the pagespeed-memcached and the pagespeed-filecache interfaces.  If there were problems with our caching implementation, this is one place I would expect them to show up.  This is a microbenchmark, stressing one single path through our system.

I would also like to run a more realistic load test, with more than enough resources to fill the cache and a real-life hit rate curve, but that's not what this test is going for.

(When running the more realistic test I'm not going to be able to compare to the file cache, because the file cache doesn't limit it's usage of disk space effectively under very high load.  It only checks once an hour whether it's using too much space.  But I can still compare our memcached and redis implementations.)

**Comment by jeffkaufman** [comment]
@jmarantz "I think Jeff's benchmark won't do many writes."

Right: it's not testing write performance at all, because I warmed each cache for the resource beforehand.  It's a test of read performance only.

(In a steady state with a large enough cache pagespeed should be doing way more reads than writes.  Basically, it should only be issuing writes when a resource changes.  So cache read performance should be one of the largest factors in overall performance.)

**Comment by jeffkaufman** [comment]
What I've tested so far:

* single hot file read performance (small file, large file)

What I still want to test:

* single hot file read performance on memcached with no `MultiGet`s
* write performance (small files, large files)
* combined performance (realistic mixed load of hot and cold resources, with reads and writes)
* stability: performance when external cache goes unresponsive in the middle of the test

**Comment by jeffkaufman** [comment]
Also, I'm sorry I didn't give more context initially on what my test was intended to determine!

**Comment by fcorriga** [comment]
no problems it's clear now and i think the difference in performance is due, as already said by jmarantz , in Redis Multiget

**Comment by jeffkaufman** [comment]
Results for memcached with no `MultiGet` [1]:

| tiny file | large file |
| --: | --: |
| 3466.82 | 773.76 |
| 3372.87 | 775.48 |
| 3492.76 | 781.49 |
| 3499.37 | 775.21 |
| 3500.77 | 772.61 |
| 3482.38 | 774.86 |
| 3429.39 | 770.72 | 
| 3519.39 | 773.22 |
| 3542.96 | 783.74 |
| 3509.85 | 769.82 |

Averages, and comparisons to normal memcached performance:

* tiny file: 3482 (96.5%)
* large file: 775 (106%)

This shows that MultiGet improves memcached's performance in the small-file case by 3.6%, so naively we would expect a similar improvement if we added it to our redis client.  That would move redis from ~3442 QPS to ~3566 QPS, moving it from 95% of memcached's performance to 99% of memcached's performance.  My numbers are somewhat noisy, so differences of 1% probably aren't meaningful.

On the other hand, this shows MultiGet makes memcached's performance worse in the large file case.  This makes sense, because large responses benefit very little from batching and the batching has some overhead.   We don't know, at the time we do our lookups, whether the response is going to be large or small, so we can't just not batch requests that are going to get large responses.  So some combination of our redis implementation and redis is 5% worse than with memcached in single-file reads.

And on the gripping hand, these results aren't as trustworthy as I would like them to be, because (a) they weren't interleaved with the earlier ones and (b) due to a power outage my machine restarted between these runs and the earlier ones.

[1] This is with `AprMemCache::MultiGet` simply removed, so it will fall back to repeated calls to `Get`.

**Comment by jeffkaufman** [comment]
Testing read+write performance.  In this test, we have a very large number of urls which we each only hit once.  Same config as before, with the addition of:

```
  ModPagespeedInPlaceWaitForOptimized on
  ModPagespeedInPlaceRewriteDeadlineMs 5000
```

This will make us wait for each resource until it's optimized.  For each one, we will do something like a metadata cache lookup [miss], then an http cache lookup [miss], then a disk read, then an http cache write, then a metadata cache write.

  Using a slightly tweaked `siege_instant_ipro.sh`:

|redis|memcached|file|
| ---: | ---: | ---: |
|2485.09|2466.70|2294.10|
|2460.63|2470.36|2120.44|
|2511.93|2498.75|2265.01|
|2479.54|2508.15|2250.73|
|2485.09|2480.77|2250.23|
|2490.04|2446.78|2259.89|
|2474.02|2467.31|2232.14|
|2466.09|2480.77|2190.10|
|2463.66|2486.94|2235.64|
|2416.04|2472.19|2243.16|

Averages:

* redis: 2473 (111%)
* memcached: 2478 (111%)
* file: 2234 (100%)

![tmp](https://cloud.githubusercontent.com/assets/262566/18726816/23905066-8014-11e6-9c74-a844b0c7e9ce.png)

(`siege_instant_ipro.sh` normally has concurrency=10, but for comparision with the results above I'm running with concurrency=50)

**Comment by jeffkaufman** [comment]
(in all the images, "left vertical axis title" should be "http requests served per second".)

**Comment by jeffkaufman** [comment]
More testing, now with a more realistic load pattern.  One large jpeg file, fetched 20k times with query params to make PageSpeed treat it as many files.  Query params follow a load pattern taken from a real site (www.jefftk.com/hits-frequency.txt.gz) Unlike in the writing test above, no `WaitForOptimized` setting.  Redis and Memcached are still at 64MB of ram, small enough that they can't close to keep up with what PageSpeed wants from them.

Tracking both QPS and %Optimized:

|redis-qps|redis-opt%|memcached-qps|memcached-opt%|
| ---: | ---: | ---: | ---: |
|438.50 |38.85%|437.83|27.13%|
|445.34|38.89%|427.81|26.72%|
|452.39|38.40%|444.74|28.50%|

I'm not sure why redis would have such a higher fraction of resources optimized, and this may be a problem in my test environment.  Don't trust these numbers yet!  (I'm also not running a full ten runs, since they're so clearly far apart after just three.)

**Comment by fcorriga** [comment]
just out of curiosity, are you using persistent connections with Redis?

**Comment by yeputons** [comment]
@fcorriga Yes, each mod_pagespeed instance (i.,e. one per apache child process) establishes single connection to Redis on start up and uses it for all requests.

**Comment by fcorriga** [comment]
uhm i wonder if you implemented pipelining too at this early stage

**Comment by yeputons** [comment]
@fcorriga We did not. It seems to me that pipelining should not help much when network latency between mod_pagespeed and Redis is negligible (i.e. they're on the same machine), should it?

**Comment by fcorriga** [comment]
ok, it was just a thought because you're connecting to redis thru the network stack and not thru unix socket, so there is some latency and network overhead.

**Comment by jeffkaufman** [comment]
@cheesy thought maybe the redis vs memcached difference was cache batcher (multiget) related, so here's a run with the cache batcher disabled (removing the appropriate lines in `SystemCaches::ConstructExternalCacheInterfacesFromBlocking`:

|redis-qps|redis-opt%|memcached-qps|memcached-opt%|
| ---: | ---: | ---: | ---: |
|448.83|38.95%|429.46|27.08%|

So, it's not that.

**Comment by jeffkaufman** [comment]
Looks like the problem was that the redis tests I just ran weren't hitting the redis server I had manually enabled, but a different redis server I had running in the background from running tests before.  ::shame::

Starting a new round of tests now.

**Comment by jeffkaufman** [comment]
With my redis server fixed, here are the new results:

|redis-qps|redis-opt%|memcached-qps|memcached-opt%|
| ---: | ---: | ---: | ---: |
|399.92|18.45%|447.23|27.50%|
|429.09|17.47%|443.07|28.14%|
|419.11|17.31%|447.63|28.58%|

These are puzzling in the opposite direction: I'm not sure why redis is giving such a low optimization fraction.

Looking at DB stats, I see:

```
$1737
# Server
redis_version:2.8.4
redis_git_sha1:00000000
redis_git_dirty:0
redis_build_id:a44a05d76f06a5d9
redis_mode:standalone
os:Linux 3.13.0-95-generic x86_64
arch_bits:64
multiplexing_api:epoll
gcc_version:4.8.2
process_id:26283
run_id:93abccd770a79103d0b7cce91fb602b1939d1227
tcp_port:6380
uptime_in_seconds:73
uptime_in_days:0
hz:10
lru_clock:656511
config_file:

# Clients
connected_clients:7
client_longest_output_list:0
client_biggest_input_buf:0
blocked_clients:0

# Memory
used_memory:57280216
used_memory_human:54.63M
used_memory_rss:59486208
used_memory_peak:67241736
used_memory_peak_human:64.13M
used_memory_lua:33792
mem_fragmentation_ratio:1.04
mem_allocator:jemalloc-3.4.1

# Persistence
loading:0
rdb_changes_since_last_save:3628
rdb_bgsave_in_progress:0
rdb_last_save_time:1474571438
rdb_last_bgsave_status:ok
rdb_last_bgsave_time_sec:-1
rdb_current_bgsave_time_sec:-1
aof_enabled:0
aof_rewrite_in_progress:0
aof_rewrite_scheduled:0
aof_last_rewrite_time_sec:-1
aof_current_rewrite_time_sec:-1
aof_last_bgrewrite_status:ok

# Stats
total_connections_received:7
total_commands_processed:49019
instantaneous_ops_per_sec:0
rejected_connections:0
sync_full:0
sync_partial_ok:0
sync_partial_err:0
expired_keys:0
evicted_keys:3198
keyspace_hits:14658
keyspace_misses:30727
pubsub_channels:0
pubsub_patterns:0
latest_fork_usec:0

# Replication
role:master
connected_slaves:0
master_repl_offset:0
repl_backlog_active:0
repl_backlog_size:1048576
repl_backlog_first_byte_offset:0
repl_backlog_histlen:0

# CPU
used_cpu_sys:2.72
used_cpu_user:0.44
used_cpu_sys_children:0.00
used_cpu_user_children:0.00

# Keyspace
db0:keys=234,expires=0,avg_ttl=0
```

And memcached:
```
stats
STAT pid 23303
STAT uptime 348
STAT time 1474571632
STAT version 1.4.14 (Ubuntu)
STAT libevent 2.0.21-stable
STAT pointer_size 64
STAT rusage_user 0.290310
STAT rusage_system 2.839596
STAT curr_connections 5
STAT total_connections 12
STAT connection_structures 11
STAT reserved_fds 20
STAT cmd_get 47459
STAT cmd_set 3353
STAT cmd_flush 0
STAT cmd_touch 0
STAT get_hits 20705
STAT get_misses 26754
STAT delete_misses 0
STAT delete_hits 0
STAT incr_misses 0
STAT incr_hits 0
STAT decr_misses 0
STAT decr_hits 0
STAT cas_misses 0
STAT cas_hits 0
STAT cas_badval 0
STAT touch_hits 0
STAT touch_misses 0
STAT auth_cmds 0
STAT auth_errors 0
STAT bytes_read 914024889
STAT bytes_written 4727531667
STAT limit_maxbytes 67108864
STAT accepting_conns 1
STAT listen_disabled_num 0
STAT threads 4
STAT conn_yields 0
STAT hash_power_level 16
STAT hash_bytes 524288
STAT hash_is_expanding 0
STAT expired_unfetched 0
STAT evicted_unfetched 863
STAT bytes 52037886
STAT curr_items 1993
STAT total_items 3353
STAT evictions 1040
STAT reclaimed 0
```

Some differences:
* Redis evicted 3,198 keys while Memcached evicted 1,040.
* Redis currently contains 234 items while Memcached contains 1,993
* Redis had 14,658 hits to Memcached's 20,705
* Redis had 30,727 misses to Memcached's 26,754

I'm not entirely sure why Redis is performing worse.  One possibility, hinted at by Memcached having more items and more transactions (sum of hits and misses) is that maybe Memcached is doing a better job of preserving metadata cache entries, the tiny bits of data that tell us what url goes with what optimized image?  This is less a question of "which cache is better" and more "which cache is better suited for pagespeed, at least in it's default configuration".  It's also possible that Redis just doesn't do well at very small cache sizes, and I would see something different if I ran a much larger test (more files, GB+ cache, more time).

I'm going to stop the speed benchmarking here, at least for the near future.  Redis works, right now I'd guess it's not quite as good as memcached for our usage in cache-too-small situations, but not critically so.

I'm going to do a couple stability tests, and then move on to fixing other things for the upcoming release.

**Comment by jeffkaufman** [comment]
Stability test: same test as I just did, but with redis cache being very flaky.  Both suspending/restoring it at random times and killing/restarting it.  PageSpeed+Redis maintained 346.02 QPS with an optimization rate of 12.32% while PageSpeed+Memcached got 411.52 QPS with an optimization fraction of 1.48%.

No crashes or other instability. (And no crashes during all the tests above either.)

The reason Redis ges a higher optimization percentage here is probably that our client has backoff logic to handle unhealthy connections, while our Memcached server just gives up and says "I'll try again in 30s".

**Comment by jeffkaufman** [comment]
One more quick test: @hillsp 's central controller (off by default) changes how we interact with the cache.  I was curious if this affected our optimization percentage across the two servers.  Here's the performance test from above, with the central controller enabled:

|redis-qps|redis-opt%|memcached-qps|memcached-opt%|
| ---: | ---: | ---: | ---: |
|419.90|19.21|441.01|28.68%|
|429.92|18.53%|446.93|26.94|

No real difference.

**Comment by jeffkaufman** [comment]
No more updates for a while here; happy to answer questions though.

**Comment by yeputons** [comment]
@jeffkaufman If both Redis and Memcached are limited by amount of memory only (not some other metric), I suspect that Redis may use more memory per entry, so it's unable to store as much keys as memcached. This may result in higher eviction rates and lower qps.

**Comment by jeffkaufman** [comment]
@yeputons If redis needs more memory per entry, then that would definitely explain the difference I measured.

**Comment by jeffkaufman** [comment]
@yeputons Separately, @hillsp mentioned that you were worried we would do the wrong thing if keys contained curly braces or spaces, since redis would interpret them.  I wrote a test for this, however, and it seems like it's fine?

```
// Make sure curly braces in keys aren't treated specially.                      
TEST_F(RedisCacheTest, CurlyBraces) {
  if (!InitRedisOrSkip()) {
    return;
  }
  CheckPut("{1}NameA", "Value1A");
  CheckPut("{2}NameB", "Value2B");
  CheckPut("{2}NameC", "Value2C");

  CheckGet("{1}NameA", "Value1A");
  CheckGet("{2}NameB", "Value2B");
  CheckGet("{2}NameC", "Value2C");
}

// And spaces                                                                    
TEST_F(RedisCacheTest, Spaces) {
  if (!InitRedisOrSkip()) {
    return;
  }
  CheckPut("1 NameA", "Value1A");
  CheckPut("2 NameB", "Value2B");
  CheckPut("2 NameC", "Value2C");

  CheckGet("1 NameA", "Value1A");
  CheckGet("2 NameB", "Value2B");
  CheckGet("2 NameC", "Value2C");
}
```

**Comment by yeputons** [comment]
@jeffkaufman spaces could be a problem (they are not as hiredis is smart enough), curly braces are problem in case of Redis Cluster, as they specify "<a href="http://redis.io/topics/cluster-spec#keys-hash-tags">hash tags</a>" and skew distribution of keys from "random based on key name" to "random on what is inside curly brackets".

**Comment by fcorriga** [comment]
redis_version:2.8.4
Please update to 3.2.3 ...

**Comment by f4rsh4d** [comment]
Actually lots of improvement happened to version 3.0 and 3.2 , specially in clustering part of redis .

**Comment by jeffkaufman** [comment]
With https://github.com/pagespeed/mod_pagespeed/commit/75fb7f5a261674c0107b25cd97f4d6f530e601e1 we now have clustering support

**Comment by f4rsh4d** [comment]
@jeffkaufman thanks for the hard work , i'll test it today or tomorrow ,
is it safe to use it on a production env ?

**Comment by jeffkaufman** [comment]
@f4rsh4d "is it safe to use it on a production env ?"

I'm going to do some stress testing and benchmarking this morning, which you might want to wait for.

**Comment by fcorriga** [comment]
@jeffkaufman: please update redis

**Comment by jeffkaufman** [comment]
@fcorriga Thanks!  I'm going to be running these tests with 3.2.0.

**Comment by fcorriga** [comment]
@jeffkaufman: 3.2.0 has some bugs. 3.2.4 is stable

**Comment by jeffkaufman** [comment]
@fcorriga How severe bugs?  In my test setup it's far easier to test with 3.2.0; I'd have to build 3.2.4 from source.

**Comment by jeffkaufman** [comment]
Looks like I need a followup to 75fb7f5 since our config parser still thinks redis cluster isn't supported ::shame::

**Comment by fcorriga** [comment]
@jeffkaufan: if you use debian or ubuntu there are pre-built packages in 3rd party repositories. if you use arch, just add the community repository and get it from there.

**Comment by jeffkaufman** [comment]
> Looks like I need a followup to 75fb7f5 since our config parser still thinks redis cluster isn't supported ::shame::

Actually, I was just starting it wrong plus I forgot to update an error message.  We only expect you to specify one node from the cluster and I was trying to specify all of them.

**Comment by jeffkaufman** [comment]
Testing with a large image, 40k hits.

Redis cluster is three masters and three slaves with 64m each (possibly these should be 32?).  Memcached and non-cluster redis each have 192m.  Redis version is now 3.2.0.

Performance:

|cluster|redis|memcached|
| ---: | ---: | ---: |
|428.95|430.52|436.11|
|432.67|429.78|438.55|
|430.62|397.85|432.81|
|438.02|421.63|437.73|
|432.29|430.52|436.82|

Optimized fraction:

|cluster|redis|memcached|
| ---: | ---: | ---: |
|35.62%|31.25%|27.67%|
|35.98%|29.15%|28.63%|
|37.25%|29.58%|27.66%|
|36.08%|27.55%|27.74%|
|35.84%|30.08%|25.74%|

**Comment by jeffkaufman** [comment]
Redis is ready to go out in the next release.  Writing doc now.

Closing this issue, since it's full implemented on master.

**Comment by onnimonni** [comment]
I couldn't find any docs about the configuration yet. Will the docs will be available over here? -> https://developers.google.com/speed/pagespeed/module/system

**Comment by yeputons** [comment]
@onnimonni I believe it will appear there when the next release is published.

Short version: you can _experiment_ by using master branch of mod_pagespeed and adding `ModPagespeedRedisServer localhost:6379` directive to Apache config (`pagespeed RedisServer localhost:6379` for nginx), which should fully enable Redis (you can check it on admin site by looking at "Cache Structure"). If you're using Redis Cluster, you can point `RedisServer` to any of nodes and mod_pagespeed should discover remaining nodes automatically. In any case, you should configure Redis <a href="http://redis.io/topics/lru-cache">as LRU cache</a> (`allkeys-lru` or `allkeys-random`).

**Comment by onnimonni** [comment]
Excellent! Is it possible to use redis backend with authentication? If so how do I provide the password for nginx?

**Comment by yeputons** [comment]
@onnimonni Not yet, unfortunately. Are you referring to Redis' `requirepass` directive and corresponding <a href="http://redis.io/commands/AUTH">AUTH command</a> which sends instance-dependent password to Redis in plaintext?

**Comment by jmarantz** [comment]
Just to clarify, this code is checked into trunk and is in pretty good shape now.  However, we have not made a formal release with this yet, due to unrelated issues with some of our dependencies and build infrastructure.  Hopefully soon.

**Comment by onnimonni** [comment]
@yeputons exactly! I hope this feature will be added later on. Thanks again 👍

**Comment by jeffwidman** [comment]
> We only expect you to specify one node from the cluster and I was trying to specify all of them.

In production, I'd prefer that any client config allow specifying 2-3 nodes of a cluster so that if the first one is down, but the rest of the cluster is up, then the client can just go check the others until it finds one that's alive. Otherwise, losing your "bootstrap" node means that on the next restart of pagespeed the redis clsuter will be unlocatable.

**Comment by jeffkaufman** [comment]
> check the others until it finds one that's alive

Yes, that would be better, and is on our list of ways to improve our Redis integration, but it's not going to make it into the next release.

**Comment by ciarans** [comment]
Hey guys, is there any news regarding the implementation of the AUTH command?

**Comment by jeffkaufman** [comment]
@ciarans no one has been working on that, sorry!

If you or anyone else wants to have a go at writing a pull request I'd be happy to review it.  See https://github.com/pagespeed/mod_pagespeed/wiki/Development for getting started.

**Comment by jmarantz** [comment]
@ciarans I'd recommend also reporting this as a new bug, rather than
appending to this older, closed bug.

-Josh
